### ISSUE TITLE
MemoryGrainStorage does not work with State types that do not have a codec

### ISSUE THREAD

**Issue by koenbeuk** [dotnet/orleans#8101]
While upgrading a 3.6.2 codebase to Orleans 7 RC2, we noticed our tests are failing with exceptions similar to these: `Orleans.Serialization.CodecNotFoundException: Could not find a copier for type Linker.Core.Github.Grains.GithubRepoDiscoveries.GithubRepoDiscoveryGrainState.`

[Memory storage](https://github.com/dotnet/orleans/blob/main/src/Orleans.Persistence.Memory/Storage/MemoryStorage.cs) works by storing the State of a grain in a dedicated grain. This means that the state type needs to have a serializer/copier/activator. When targeting (real) persistent grain storage providers, the storage provider will take care of the serialization (e.g. JSON).

This becomes an obvious issue starting from Orleans 7 as before, a serializer would be generated for you during runtime.

**Comment by koenbeuk** [comment]
My proposal: we make MemoryGrainStorage flexible so that we can configure a Serializer strategy, usage could look like this:

```csharp
siloBuilder.AddMemoryGrainStorageAsDefault(options => {
  options.UseSerializer<SystemTextJsonSerializer>();
});
```

Now MemoryGrainStorage will json serialize grain state before forwarding the serialized content to the MemoryStorageGrain.

Happy to PR this after feedback.

**Comment by ReubenBond** [comment]
Hi @koenbeuk, thanks for pointing this out. I agree with you. `MemoryGrainStorage` should follow the same pattern as the other storage providers by implementing `IStorageProviderSerializerOptions`.

See these references:
* Implement the interface: https://github.com/dotnet/orleans/blob/fb5ce38be59844ab3c42b2e83d302a6578645609/src/Azure/Orleans.Persistence.AzureStorage/Providers/Storage/AzureTableStorageOptions.cs#L12
* Set the default: https://github.com/dotnet/orleans/blob/1c0ee4370d540547e6070c4caa71c50baf45d441/src/Azure/Orleans.Persistence.AzureStorage/Hosting/AzureTableSiloBuilderExtensions.cs#L53

A PR would be very welcome, thank you!

**Comment by shoneefd** [comment]
Hi @ReubenBond, I have some fix that I think will fix this issue, but I've not done work in this project before. Is there documentation on how I can run the testing suite and what the procedure is for verifying my code works? Thanks!

**Comment by ReubenBond** [comment]
You can open a PR and let the CI run tests. You can also run tests from the command line using `dotnet test --filter "Category=BVT"`, I believe. You can run BVT tests in VS, too

**Comment by ReubenBond** [comment]
Fixed by #8250
